### ISSUE TITLE
Content Security Policy: unsafe-eval

### ISSUE THREAD

**Issue by brianmhunt** [WebReflection/backtick-template#1]
It'd be quite valuable if there were a version of this that did not use `Function` or `eval`, so it could be used with CSP.

:beers:

**Comment by WebReflection** [comment]
It's impossible to avoid the usage of `Function`, which is not the equivalent of `eval`, simply because the `with` statement wouldn't be surely available otherwise.

`Function` is indeed the only way to opt out from `"use strict";` and without `Function` and the `with` statement this utility has no reason to exist.

**Comment by WebReflection** [comment]
Actually, it looks like on the global scope `"use strict"` cannot be enabled so I might try to create a version CSP friendly like I've done for [common-js](https://github.com/WebReflection/common-js/blob/master/csp.html)

**Comment by WebReflection** [comment]
OK, using a script with a runtime injected `textContent` has very poor performance and it's not any more secure.

I'm afraid you need to enable `unsafe-eval` which IMO in a CSP environment is not too bad, if you manage to inject undesired code it means already you're accepting random input from code that sneaked in already, right?

https://blogs.dropbox.com/tech/2015/09/unsafe-inline-and-nonce-deployment/

**Comment by brianmhunt** [comment]
Thanks.

I've written a substitute for "Function" (basically a javascript lambda-parser) in [tko.provider](/knockout/tko.provider) for the Knockout.js project's next version.  Looking through our history, you'll see we had to deal with the same (`with($context){with($data){...}}`).  Its performance is excellent.

That's probably overkill for this project, and as you say it's not necessarily that much more restrictive, but certainly if you were keen on supporting CSP with `unsafe-eval` you'd probably need a parser-like system something like tko.provider.

Another way to isolate the code is by putting the template rendering into a web worker, basically a sandbox that doesn't have `unsafe-eval`, so it won't have direct access to the DOM, among other restrictions.

Just food for thought! 🍕

**Comment by WebReflection** [comment]
I think a whole JS parser for a string template would basically make this whole concept, and its simplicity, performance, and file size, impractical and useless.

Like I've said, if you are in CSP there's no way a script can inject an external URL you don't trust already and the template strings are static, not something that comes from the user input, right?

Even though, it would be `'hello ${user.input}'` and no matter if the user used malicious code, all he'll get is a string with such code in it, nothing is evaluated.

As summary, CSP concerns for this utility are overrated because even with a whole JS parser that doesn't use Function, if you do `template(user.input)` directly you gonna have every other problem as well.

Basically, ending up parsing and enabling code to work, is at the end of the day what Function does already: is no more secure, just slower, IMO.

So, no direct access to the DOM, you write code, you have static strings like it is for regular `ES6` strings ... your concern is basically that one would write
```js
const disaster = `[${eval(user.input)}]`
```

but that's either exactly what you need, and you know the risks, or a non real-world scenario, and you just enable unsafe-eval and be sure in the whole projects you only use template with statically defined strings and not defined through user input like `template(user.input)`.

Thanks for the link to the project thought, its interesting stuff regardless this ticket 👍
